libbpf-rs/src/query.rs

```diff
@@ -498,29 +498,80 @@ gen_info_impl!(
 // TODO: Document members.
 #[allow(missing_docs)]
 pub struct BtfInfo {
-    pub btf: u64,
-    pub btf_size: u32,
+    pub name: String,
+    pub btf: Vec<u8>,
     pub id: u32,
 }
 
 impl BtfInfo {
-    fn from_uapi(_fd: i32, s: libbpf_sys::bpf_btf_info) -> Option<Self> {
-        Some(Self {
-            btf: s.btf,
-            btf_size: s.btf_size,
-            id: s.id,
-        })
+    fn load_from_fd(fd: i32) -> Option<Self> {
+        let mut item: libbpf_sys::bpf_btf_info = unsafe { std::mem::zeroed() };
+        let mut btf: Vec<u8> = Vec::new();
+        let mut name: Vec<u8> = Vec::new();
+
+        let item_ptr: *mut libbpf_sys::bpf_btf_info = &mut item;
+        let mut len = size_of::<libbpf_sys::bpf_btf_info>() as u32;
+
+        let ret =
+            unsafe { libbpf_sys::bpf_obj_get_info_by_fd(fd, item_ptr as *mut c_void, &mut len) };
+        if ret != 0 {
+            return None;
+        }
+
+        // The API gives you the ascii string length while expecting
+        // you to give it back space for a nul-terminator
+        item.name_len += 1;
+        name.resize(item.name_len as usize, 0u8);
+        item.name = name.as_mut_ptr() as *mut c_void as u64;
+
+        btf.resize(item.btf_size as usize, 0u8);
+        item.btf = btf.as_mut_ptr() as *mut c_void as u64;
+
+        let ret =
+            unsafe { libbpf_sys::bpf_obj_get_info_by_fd(fd, item_ptr as *mut c_void, &mut len) };
+        if ret == 0 {
+            Some(BtfInfo {
+                name: String::from_utf8(name).unwrap_or_else(|_| "(?)".to_string()),
```

Comment:
One more suggestion: how about we make `BtfInfo::name` a `CString`. This way we can forego this opinionated fallible conversion and let users choose how to handle errors and it's barely more than a `to_string_lossy` call on their end if they opt not to.
It's what we do in the existing BTF functionality: https://docs.rs/libbpf-rs/0.21.2/libbpf_rs/btf/struct.BtfType.html#method.name

Reply:
For what it's worth, ProgramInfo::name is a String as well. Should that be changed?

Reply:
Ah, I missed that. Yes, that would be great. Though I won't insist you do it as you didn't really touch that member. So I'll leave it up to you. I added it to my list of things to do as well.